# Challenge

We are the team composed by:

- Carlos Aznar (capitán)
- Jason Carneiro
- Elías Carrilo
- Alejandro Rodríguez
- Adrián Blázquez

In this project we discuss how to make three models accurate and precise (one geometric, a decision tree and a GaussianNB). To do so:

# 1. The knowledge domain is analyzed
# 2. Exploratory analysis
# 3. Machine learning model
# 4. Result of the competition
# 5. Presentation discussion


# 1. The knowledge domain is analyzed

Analice el dataset proporcionado, entendiendo las variables del mismo, y su significado. Discuta en grupo sus intuiciones sobre qué variables pueden ser buenos predictores del riesgo asociado. Resuman sus conclusiones en el cuaderno (notebook) de Jupyter.

Para ello se lee el dataset de train para cargar el data frame.

In [1]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

df = pd.read_csv('data/train.csv')
df

id  LoanNr_ChkDgt                            Name  \
0      d30838cf806     6340864010  QUALITY DOCUMENT SOLUTIONS INC   
1      14560b856a6     1670916004  ATLANTIC INDUSTRIAL WOOD PRODU   
2      051e5c3899f     7568313003              C J TREEHOUSE INC.   
3      eb223b05ea8     4585814005                        WINGSTOP   
4      c80c89ac964     2003866003      H & H MACHINE & SUPPLY INC   
...            ...            ...                             ...   
49315  7bc068859e5     4999154008              SUPREME BAKERY INC   
49316  1f0420db159     1401875000              Yunmi Sushi, Corp.   
49317  7f4cbcda5f5     5823093008          SOUTHSIDE ACE HARDWARE   
49318  4dcd80b2311     2713645005     THE FORSBERG LAW FIRM, P.C.   
49319  07e1e2837e5     9391233006  BOX BROTHERS ENTERPRISES, INC.   

                City State                            Bank BankState  \
0             ODESSA    TX     WELLS FARGO BANK NATL ASSOC        SD   
1             KELLER    TX     WELLS FARGO BANK NATL ASSOC        CA   
2            EL PASO    TX      BANK OF AMERICA NATL ASSOC        TX   
3        SAN ANTONIO    TX                    COMPASS BANK        AL   
4             ODESSA    TX                  SOUTHWEST BANK        TX   
...              ...   ...                             ...       ...   
49315        EL PASO    TX  JPMORGAN CHASE BANK NATL ASSOC        IL   
49316        HOUSTON    TX  JPMORGAN CHASE BANK NATL ASSOC        IL   
49317  GRAND PRAIRIE    TX                BANK OF THE WEST        TX   
49318     MONTGOMERY    TX          CAPITAL ONE NATL ASSOC        VA   
49319        HOUSTON    TX     WELLS FARGO BANK NATL ASSOC        SD   

      ApprovalDate ApprovalFY  NoEmp  ...  RetainedJob  FranchiseCode  \
0        22-May-03       2003     10  ...           10              1   
1         7-Apr-06       2006      1  ...            1              1   
2        12-Sep-94       1994      5  ...            0              1   
3        18-Jun-01       2001      1  ...            1          89769   
4        12-May-06       2006      5  ...            5              1   
...            ...        ...    ...  ...          ...            ...   
49315    23-Oct-01       2002     13  ...           13              1   
49316     4-Aug-05       2005      1  ...            1              0   
49317     3-Aug-93       1993      8  ...            0           1560   
49318    21-Sep-07       2007      3  ...            3              0   
49319     9-Jul-96       1996      7  ...            0              1   

       UrbanRural  RevLineCr  LowDoc ChgOffDate DisbursementDate  \
0               1          0       N        NaN        30-Jun-03   
1               1          Y       N        NaN        31-Jul-06   
2               0          N       Y        NaN        31-Oct-94   
3               1          0       N        NaN        31-Aug-01   
4               1          0       N        NaN        31-May-06   
...           ...        ...     ...        ...              ...   
49315           0          0       N        NaN        31-Mar-02   
49316           1          Y       N        NaN        30-Sep-05   
49317           0          N       N        NaN        31-Jan-94   
49318           1          N       N        NaN        31-Oct-07   
49319           0          0       N        NaN        31-Dec-96   

      DisbursementGross BalanceGross Accept  
0          $150,000.00        $0.00       1  
1          $250,000.00        $0.00       1  
2           $53,000.00        $0.00       1  
3          $170,000.00        $0.00       1  
4          $121,000.00        $0.00       1  
...                 ...          ...    ...  
49315       $63,000.00        $0.00       1  
49316       $25,000.00        $0.00       1  
49317      $200,000.00        $0.00       1  
49318       $35,000.00        $0.00       1  
49319       $90,000.00        $0.00       1  

[49320 rows x 22 columns]

In [2]:
# we can get the number of samples and features
df.shape

(49320, 22)

In [3]:
# Information about columns and their types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49320 entries, 0 to 49319
Data columns (total 22 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 49320 non-null  object 
 1   LoanNr_ChkDgt      49320 non-null  int64  
 2   Name               49319 non-null  object 
 3   City               49320 non-null  object 
 4   State              49320 non-null  object 
 5   Bank               49120 non-null  object 
 6   BankState          49120 non-null  object 
 7   ApprovalDate       49320 non-null  object 
 8   ApprovalFY         49320 non-null  object 
 9   NoEmp              49320 non-null  int64  
 10  NewExist           49313 non-null  float64
 11  CreateJob          49320 non-null  int64  
 12  RetainedJob        49320 non-null  int64  
 13  FranchiseCode      49320 non-null  int64  
 14  UrbanRural         49320 non-null  int64  
 15  RevLineCr          49031 non-null  object 
 16  LowDoc             491

In [4]:
# Basic statistics of the dataset in all the numeric columns
df.describe()

LoanNr_ChkDgt         NoEmp      NewExist     CreateJob   RetainedJob  \
count   4.932000e+04  49320.000000  49313.000000  49320.000000  49320.000000   
mean    4.941161e+09     10.794505      1.317361      9.871006     11.930211   
std     2.530125e+09     44.645641      0.467497    262.815556    266.334776   
min     1.000254e+09      0.000000      0.000000      0.000000      0.000000   
25%     2.731376e+09      2.000000      1.000000      0.000000      0.000000   
50%     4.598009e+09      4.000000      1.000000      0.000000      0.000000   
75%     7.141562e+09     10.000000      2.000000      1.000000      3.000000   
max     9.995563e+09   4000.000000      2.000000   8800.000000   9500.000000   

       FranchiseCode    UrbanRural        Accept  
count   49320.000000  49320.000000  49320.000000  
mean     3933.928062      0.658942      0.810929  
std     15141.332203      0.595339      0.391569  
min         0.000000      0.000000      0.000000  
25%         1.000000      0.000000      1.000000  
50%         1.000000      1.000000      1.000000  
75%         1.000000      1.000000      1.000000  
max     92000.000000      2.000000      1.000000

 Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 49320 non-null  object 
 1   LoanNr_ChkDgt      49320 non-null  int64  
 2   Name               49319 non-null  object 
 3   City               49320 non-null  object 
 4   State              49320 non-null  object 
 5   Bank               49120 non-null  object 
 6   BankState          49120 non-null  object 
 7   ApprovalDate       49320 non-null  object 
 8   ApprovalFY         49320 non-null  object 
 9   NoEmp              49320 non-null  int64  
 10  NewExist           49313 non-null  float64
 11  CreateJob          49320 non-null  int64  
 12  RetainedJob        49320 non-null  int64  
 13  FranchiseCode      49320 non-null  int64  
 14  UrbanRural         49320 non-null  int64  
 15  RevLineCr          49031 non-null  object 
 16  LowDoc             49160 non-null  object 
 17  ChgOffDate         9571 non-null   object 
 18  DisbursementDate   49194 non-null  object 
 19  DisbursementGross  49320 non-null  object 
 20  BalanceGross       49320 non-null  object 
 21  Accept             49320 non-null  int64

# 2. Exploratory analysis

Analice el dataset para confirmar (o desestimar) sus hipótesis y discuta sus descubrimientos en el cuaderno.

# 3. Machine learning model

Desarrolle uno o más modelos de aprendizaje automático para clasificar si una PYME debe (o no) recibir un crédito. Debe probar al menos: un algoritmo geométrico, un árbol de decisión y el algoritmo secreto recibido por el grupo. Discuta los resultados y valídelos.

## 3.1. Geometric model

## 3.2. Decision tree

## 3.3. GaussianNB

# 4. Result of the competition

Una vez disponga de un modelo, deberá usar la plataforma Kaggle para probar su modelo con datos, que combinan datos nuevos con los datos existentes. Puede realizar tantos intentos como desee. Razone los resultados en el cuaderno.

# 5. Presentation discussion